### PR TITLE
Introduce APP_ENV and remove RACK_ENV

### PR DESCRIPTION
**Pull request by dmathieu**
`RACK_ENV` should not be used with values other than `development` and `deployment`. Therefore, it can't be used to check for `production environment.

See http://www.hezmatt.org/~mpalmer/blog/2013/10/13/rack_env-its-not-for-you.html

In fact, it can even cause some issues if your app is using unicorn, as they're including internal rack middlewares relying on `RACK_ENV` to have the `deployment` value.
See https://github.com/defunkt/unicorn/blob/master/lib/unicorn.rb#L56-L79

Rack itself does that too.
https://github.com/rack/rack/blob/4e4ab39b0508aa3e59f5d7e53696ef6ae7c220ed/lib/rack/server.rb#L228

This removes using `RACK_ENV`, and replaces it with `SINATRA_ENV`.
Since it's a major changes, it should be merged in the Sinatra 2.0 branch. I'm not seeing anything of the kind, so feel free to stall this until there's discussion about it.

### CODE REVIEW

File: lib/sinatra/base.rb [hunk 1]
```diff
@@ -1834,7 +1834,7 @@ def self.force_encoding(data, *) data end
 
     reset!
 
-    set :environment, (ENV['RACK_ENV'] || :development).to_sym
+    set :environment, (ENV['SINATRA_ENV'] || :development).to_sym
```
Comment (gshutler) [inline lib/sinatra/base.rb]: I agree with the idea of this change but perhaps there's a middle ground as this would currently be a significant breaking change as it stands:

```ruby
set :environment, (ENV['SINATRA_ENV'] || ENV['RACK_ENV'] || :development).to_sym
```

Then the use of `RACK_ENV` could be deprecated and checking it removed in the next major version.

Comment (gshutler) [inline lib/sinatra/base.rb]: Sorry, I missed your note about this being for 2.0. Perhaps my suggestion would enable this to be merged to the main branch?

Comment (dmathieu) [inline lib/sinatra/base.rb]: I think enabling SINATRA_ENV, with a deprecation message when falling back to RACK_ENV, and then entirely disabling it in 2.0 makes sense.

**Comment by zzak** [comment]
Sorry you are hitting this bug.. thank you for the thorough investigation! <3

**Comment by jeremy** [comment]
Changing to `SINATRA_ENV` is legit for its own sake. However, doing it response to this blog post's reimagination of `RACK_ENV` may be premature.

TL;DR: Nobody's been doing it wrong; RACK_ENV is the Rack app environment. Rack::Server had an `-E` environment but adopted `RACK_ENV` from server & framework usage.

Discussion on Rack @ https://github.com/rack/rack/pull/831#issuecomment-89021866 and Rails @ https://github.com/rails/rails/pull/19404#issuecomment-84040161

**Comment by jeremy** [comment]
History:
  * RACK_ENV introduced in Sinatra (https://github.com/sinatra/sinatra/commit/33cbef5b6bc5f6de4b23b6255c07a147be2d11ac) in 2008, courtesy Thin (https://github.com/macournoyer/thin/commit/1e90d63f9139f6bc5dc7624ef07076bedcbb0bb3).
  * RACK_ENV introduced in Rack (https://github.com/rack/rack/commit/b433e484c9ee47ffceb7645129403df31f60db7f) in 2010 in response to established app, framework, and server convention.

**Comment by zzak** [comment]
@dmathieu Needs a rebase <3

**Comment by dmathieu** [comment]
Done :smile_cat:

**Comment by art-solopov** [comment]
:open_mouth: I didn't know this before! Apparently, my blog runs by pure coincidence.

**Comment by zzak** [comment]
@dmathieu How do you feel about this:

```ruby
# warn if RACK_ENV is set, to recommend SINATRA_ENV here, then:
set :environment, (ENV['SINATRA_ENV'] || ENV['RACK_ENV'] || :development).to_sym
```

This would allow users to choose `SINATRA_ENV` or fallback to the `RACK_ENV` if available.

I don't think we can simply ignore `RACK_ENV` but make a suggestion to enlighten people about it's proper usage.

But maybe `RACK_ENV` is always set, somewhere, so this warning will get annoying.

**Comment by mperham** [comment]
All of these pieces (Rails, Sinatra, puma, thin, sidekiq, etc) are all application infrastructure processes.  We want to determine what environment this application process is running in.  Why not `APP_ENV` so we don't have RAILS_ENV, PUMA_ENV, SINATRA_ENV variables for every piece?

**Comment by dmathieu** [comment]
@zzak that works for me. I've updated my PR.
I don't mind using APP_ENV either.

**Comment by zzak** [comment]
One thing I'll add to the discussion regarding this is, APP_ENV only seems appropriate if everyone wasn't using Rack to begin with.

We might as well just use `RACK_ENV` properly since we can safely assume anyone using Sinatra is also, and intending to, use Rack as well.

**Comment by mperham** [comment]
Happy to add it to sidekiq and I bet we could talk evanphx into adding it to puma. It should spread from there.

**Comment by gudmundur** [comment]
I'll make sure it lands in https://github.com/interagent/pliny as well.

**Comment by zzak** [comment]
@dmathieu Ok, lets change it to `APP_ENV` and see how it goes.

Worst case, we're ahead of the curve which never gains traction and we just introduce Yet Another Environment Variable to peoples apps. In which case, we do anyways if it was `SINATRA_ENV`, and still support `RACK_ENV` so I'm not of afraid of no :ghost:

**Comment by dmathieu** [comment]
LGTM. I've changed this PR to use APP_ENV instead of SINATRA_ENV.

**Comment by stefansundin** [comment]
The website still references RACK_ENV, can anyone update that? http://sinatrarb.com/configuration.html

**Comment by ioquatix** [comment]
Sorry to necro-bump this issue but after several years, have we actually reached any consensus or change on this issue?

**Comment by henrik** [comment]
@ioquatix I just came here from seeing the 2.0 release mentioning APP_ENV is now preferred, so it's listed there at any rate :) https://github.com/sinatra/sinatra/blob/master/CHANGELOG.md#200--2017-04-10

**Comment by mperham** [comment]
I've added support for APP_ENV in Sidekiq master.
